### PR TITLE
Support receptive field search of CNN models. (TPAMI paper rf-next) 

### PR DESCRIPTION
**Pull request by gasvn**
## Motivation

Merging a general receptive field search method to mmcv and mmdet. (paper: RF-Next: Efficient Receptive Field Search for Convolutional Neural Networks TPAMI 2022 [pdf](http://mftp.mmcheng.net/Papers/22TPAMI-ActionSeg.pdf))
RF-Next can improve the performance of many CNN networks on many tasks. 
more results on https://github.com/ShangHua-Gao/RF-mmdetection/tree/rfsearch/configs/rfnext

## Modification

Add rfsearch

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.
This PR relay on https://github.com/open-mmlab/mmcv/pull/2056

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

**Comment by shinya7y** [comment]
Congratulations on the PAMI acceptance!
Could the codes for ConvNeXt be cleaned based on https://github.com/open-mmlab/mmdetection/pull/7281 ?

File: configs/_base_/models/cascade_mask_rcnn_convnext_fpn.py [hunk 1]
```diff
@@ -0,0 +1,207 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+
+# All rights reserved.
+
+# This source code is licensed under the license found in the
+# LICENSE file in the root directory of this source tree.
+
+# model settings
+model = dict(
+    type='CascadeRCNN',
```
Comment (jbwang1997) [inline configs/_base_/models/cascade_mask_rcnn_convnext_fpn.py:10]: Seems the file has copyright from Meta.

File: configs/rfnext/rfsearch_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+        'decay_rate': 0.7,
+        'decay_type': 'layer_wise',
+        'num_layers': 6
+    })
+lr_config = dict(step=[27, 33])
+runner = dict(type='EpochBasedRunner', max_epochs=36)
+# runner = dict(type='EpochBasedRunnerAmp', max_epochs=36)
+
+custom_imports = dict(
+    imports=['mmcv.cnn.rfsearch'], allow_failed_imports=False)
```
Comment (jbwang1997) [inline configs/rfnext/rfsearch_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py:172]: mmcv is a necessary package of mmdetection. It has been imported directly in code. Here seems unnecessary.

File: mmdet/models/backbones/convnext.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+
+@BACKBONES.register_module()
+class ConvNeXt(BaseModule):
+    r""" ConvNeXt
+        A PyTorch impl of : `A ConvNet for the 2020s`  -
+          https://arxiv.org/pdf/2201.03545.pdf
+
+    Args:
+        in_chans (int): Number of input image channels.
```
Comment (jbwang1997) [inline mmdet/models/backbones/convnext.py:79]: After mmdet 2.25, we support ConvNeXt by using the backbone of misclassification.

**Comment by jbwang1997** [comment]
The codes in mmcv_custom need to be put into mmcv.

**Comment by BIGWangYuDong** [comment]
Unit Test all failed,  should fix UT

**Comment by gasvn** [comment]
> Unit Test all failed, should fix UT



> open-mmlab/mmcv#2056

This merge rely on the merge request from mmcv (https://github.com/open-mmlab/mmcv/pull/2056). The Unit Test failed becasue of not using the new mmcv.

**Comment by jbwang1997** [comment]
Need to delete the vim temp file configs/rfnext/search_log/convnext_cascade_maskrcnn/.local_search_config_step11.json.swp

**Comment by gasvn** [comment]
We have successfully merged the rf search operation into mmcv with the help of @zhouzaida. Now the code should be working. Is there anything we can do to contiune with this pull request?

File: mmdet/models/builder.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 import warnings
 
 from mmcv.cnn import MODELS as MMCV_MODELS
+from mmcv.cnn import RFSearchHook
 from mmcv.utils import Registry
```
Comment (jbwang1997) [inline mmdet/models/builder.py:6]:
```
try:
    from mmcv.cnn import RFSearchHook
except ImportError:
    RFSearchHook = None

File: mmdet/models/builder.py [hunk 5]
```diff
@@ -55,5 +56,20 @@ def build_detector(cfg, train_cfg=None, test_cfg=None):
...
         'test_cfg specified in both outer field and model field '
-    return DETECTORS.build(
+
+    rfsearch_cfg = cfg.pop('rfsearch_cfg', None)
+
+    detector = DETECTORS.build(
         cfg, default_args=dict(train_cfg=train_cfg, test_cfg=test_cfg))
+    if rfsearch_cfg is not None:
+        rfsearch_warp = RFSearchHook(
+            mode=rfsearch_cfg.get('mode', 'search'),
```
Comment (jbwang1997) [inline mmdet/models/builder.py:66]:
```
assert RFSearchHook is not None, 'Please install mmcv > 1.7.0'
```

File: configs/rfnext/README.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
+# RF-Next: Efficient Receptive Field Search for CNN
+
+> [RF-Next: Efficient Receptive Field Search for Convolutional Neural Networks](http://mftp.mmcheng.net/Papers/22TPAMI-ActionSeg.pdf)
+
+## Abstract
```
Comment (jbwang1997) [inline configs/rfnext/README.md:5]:
```suggestion
> [RF-Next: Efficient Receptive Field Search for Convolutional Neural Networks](http://mftp.mmcheng.net/Papers/22TPAMI-ActionSeg.pdf)

<!-- [ALGORITHM] -->

## Abstract
```

File: configs/rfnext/README.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+| RF-ConvNeXt-T | Cascade Mask R-CNN | Multiple-Branch |   3x    |  50.9   |   44.3   | [search](https://github.com/ShangHua-Gao/RF-mmdetection/tree/rfsearch/configs/rfnext/rfnext_search_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py) [retrain](https://github.com/ShangHua-Gao/RF-mmdetection/tree/rfsearch/configs/rfnext/rfnext_fixed_multi_branch_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py)  |                                                                      [model](<>) \| [log](<>)                                                                      |
```
Comment (jbwang1997) [inline configs/rfnext/README.md:18]: Please use the link of mmdetection.

Comment (lzyhha) [reply configs/rfnext/README.md:18]: Should I change it to 'https://github.com/open-mmlab/mmdetection/tree/dev/configs/rfnext/rfnext_fixed_multi_branch_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py'?

Even though this link doesn't exist yet

Comment (jbwang1997) [reply configs/rfnext/README.md:18]: You can direct write the link of master branch. The link will work after merging this pr.

File: mmdet/models/builder.py [hunk 8]
```diff
@@ -55,5 +56,20 @@ def build_detector(cfg, train_cfg=None, test_cfg=None):
...
+            config=rfsearch_cfg.get('config', None),
+            rfstructure_file=rfsearch_cfg.get('rfstructure_file', None),
+            by_epoch=rfsearch_cfg.get('by_epoch', True),
+            verbose=rfsearch_cfg.get('verbose', True),
+        )
+        detector.init_weights()
```
Comment (jbwang1997) [inline mmdet/models/builder.py:72]: Is `init_weights()` here necessary? mmdetection has initialized weights in `train.py`.

Comment (lzyhha) [reply mmdet/models/builder.py:72]: I think that it is necessary.
After rfsearch_warp.init_model(detector), the conv.weight will become conv.op_layer.weight.
So we want to init_weights before rfsearch_warp.init_model.

**Comment by jbwang1997** [comment]
Kindly ping @ZwwWayne @BIGWangYuDong @hhaAndroid @RangiLyu

**Comment by lzyhha** [comment]
Are there any progress?

**Comment by gasvn** [comment]
@ZwwWayne @BIGWangYuDong @hhaAndroid @RangiLyu @jbwang1997  Is there anything we can do to make some progress? The rf-search module has been merged to mmcv. We only need to make some small changes to make it available on mmdet.  Thanks~

**Comment by zytx121** [comment]
Hi @gasvn @lzyhha Thank you for your excellent work.
Please give us model weights by  BaiduYun or GoogleDrive. We will upload them and tell you download links.

File: configs/rfnext/README.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
+# RF-Next: Efficient Receptive Field Search for CNN
```
Comment (zytx121) [inline configs/rfnext/README.md:1]: Please add a `metafile.yml` for rfnext just like https://github.com/open-mmlab/mmdetection/blob/master/configs/solov2/metafile.yml

**Comment by lzyhha** [comment]
Thanks. I will provide the model weights and training logs in a few days.

File: mmdet/models/builder.py [hunk 10]
```diff
@@ -55,5 +60,21 @@ def build_detector(cfg, train_cfg=None, test_cfg=None):
...
+    if rfsearch_cfg is not None:
+        assert RFSearchHook is not None, 'Please install mmcv > 1.7.0'
+        rfsearch_warp = RFSearchHook(
+            mode=rfsearch_cfg.get('mode', 'search'),
+            config=rfsearch_cfg.get('config', None),
+            rfstructure_file=rfsearch_cfg.get('rfstructure_file', None),
+            by_epoch=rfsearch_cfg.get('by_epoch', True),
+            verbose=rfsearch_cfg.get('verbose', True),
+        )
+        detector.init_weights()
```
Comment (jbwang1997) [inline mmdet/models/builder.py:77]: I think it's better to put this part into train.py. In this way, we can only define `RFSearchHook` in config and delete the model `rfnext_cfg`.

BTW, it's better to separate the function of `init_model` from `RFSearchHook`. So that we don't need to build redundant `RFSearchHook` in train.py.

**Comment by jbwang1997** [comment]
Please add some UT to test the function of RFNext.

File: mmdet/models/builder.py (hunk 10 above)
Comment (lzyhha) [reply mmdet/models/builder.py:77]: By the way, init_model should be called before constructing optimizer, because there will be new parameters after calling warp_model.

Comment (lzyhha) [reply mmdet/models/builder.py:77]: Should I modify the codes of mmcv again, to separate the function of `init_model` from `RFSearchHook`?

**Comment by lzyhha** [comment]
I have removed the init_model in `builder.py` and put it in `train.py` and `test.py`. 
Are there any other issues that need to be revised?

The weights will be provided tomorrow, because I should verify the correctness of the code by re-training.
@zytx121

File: configs/rfnext/README.md (hunk 9 above)
Comment (lzyhha) [reply configs/rfnext/README.md:1]: I have added the `metafile.yaml`.

File: configs/panoptic_fpn/panoptic_fpn_r2_50_fpn_fp16_1x_coco.py [hunk 11]
```diff
@@ -0,0 +1,12 @@
+_base_ = '../panoptic_fpn/panoptic_fpn_r50_fpn_1x_coco.py'
```
Comment (jbwang1997) [inline configs/panoptic_fpn/panoptic_fpn_r2_50_fpn_fp16_1x_coco.py:1]: Seems this is a new config for panoptic_fpn. Please add the performance in panoptic_fpn README.md

Comment (jbwang1997) [inline configs/panoptic_fpn/panoptic_fpn_r2_50_fpn_fp16_1x_coco.py:1]: Can directly use _base_ = './panoptic_fpn_r50_fpn_1x_coco.py'

File: configs/rfnext/README.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+|  ConvNeXt-T   | Cascade Mask R-CNN |       NO        |   3x    |  50.4   |   43.7   |                                                                                                                                                           [config](https://github.com/facebookresearch/ConvNeXt/tree/main/object_detection)                                                                                                                                                           | [model](https://github.com/facebookresearch/ConvNeXt/tree/main/object_detection) \| [log](https://github.com/facebookresearch/ConvNeXt/tree/main/object_detection) |
```
Comment (jbwang1997) [inline configs/rfnext/README.md:18]: It's better to use the reimplement of ConvNext in mmdetection

File: configs/rfnext/README.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+| RF-ConvNeXt-T | Cascade Mask R-CNN | Multiple-Branch |   3x    |  50.9   |   44.3   | [search](https://github.com/open-mmlab/mmdetection/tree/dev/configs/rfnext/rfnext_search_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py) [retrain](https://github.com/open-mmlab/mmdetection/tree/dev/configs/rfnext/rfnext_fixed_multi_branch_cascade_mask_rcnn_convnext_tiny_patch4_window7_mstrain_480-800_giou_4conv1f_adamw_3x_coco_in1k.py)  |                                                                      [model](<>) \| [log](<>)                                                                      |
```
Comment (jbwang1997) [inline configs/rfnext/README.md:20]: Can directly use `.../tree/master/...` link here.

File: configs/rfnext/README.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+            step=0,
+            max_step=11,
+            search_interval=1,
+            exp_rate=0.5,
+            init_alphas=0.01,
+            mmin=1,
+            mmax=24,
+            num_branches=2,
+            skip_layer=[])),
+    )
```
Comment (jbwang1997) [inline configs/rfnext/README.md:77]: Need to change this part according to the new design.

File: configs/rfnext/README.md [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+
+```shell
+./tools/dist_test.sh \
+    ${rfnext_fixed_single_branch_CONFIG_FILE} \
+    ${CHECKPOINT_FILE} \
+    ${GPU_NUM} \
+    --out results.pkl \
+    --eval bbox segm \
+    --options "classwise=True"
+
```
Comment (jbwang1997) [inline configs/rfnext/README.md:145]: Seems the training and inference command are the same with other mmdetection models. We should emphasize the difference in config.

File: configs/rfnext/README.md [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+Single branch inference:
+
+```shell
+./tools/dist_test.sh \
+    ${rfnext_fixed_single_branch_CONFIG_FILE} \
+    ${CHECKPOINT_FILE} \
+    ${GPU_NUM} \
+    --out results.pkl \
+    --eval bbox segm \
+    --options "classwise=True"
```
Comment (jbwang1997) [inline configs/rfnext/README.md:144]: Seems the training and inference command are the same with other mmdetection models. We should emphasize the difference in config.

File: configs/rfnext/metafile.yml [hunk 17]
```diff
@@ -0,0 +1,232 @@
+Models:
```
Comment (jbwang1997) [inline configs/rfnext/metafile.yml:1]: Please create a `Collection` named `RF-Next`. You can refer to https://github.com/open-mmlab/mmdetection/blob/master/configs/faster_rcnn/metafile.yml#L1-L20

**Comment by lzyhha** [comment]
Hello, I have uploaded all weights and logs. [rf-mmdetection-weights(https://drive.google.com/drive/folders/1PQ47Mf9geDHMrYW4MQUb2P-yUMkvbf26?usp=sharing) @zytx121

**Comment by zytx121** [comment]
> Hello, I have uploaded all weights and logs. [rf-mmdetection-weights](https://drive.google.com/drive/folders/1PQ47Mf9geDHMrYW4MQUb2P-yUMkvbf26?usp=sharing) @zytx121

I will upload it ASAP.

**Comment by lzyhha** [comment]
log and weight of  panoptic_fpn_r2_50_fpn_fp16_1x_coco @zytx121 

[log](https://drive.google.com/file/d/11gPd23aPK2myJ3WPi48ne-VusIerkAW0/view?usp=share_link)

[weight](https://drive.google.com/file/d/1PB-3y8BSEDnp051OHgobMLAz9SfwndY5/view?usp=share_link)

**Comment by lzyhha** [comment]
I have updated the links, except panoptic_fpn_r2_50_fpn_fp16_1x_coco. @zytx121

**Comment by lzyhha** [comment]
Could you upload these two files? @zytx121

**Comment by zytx121** [comment]
@lzyhha Sorry, this upload took a little longer.
- https://download.openmmlab.com/mmdetection/v2.0/rfnext/panoptic_fpn_r2_50_fpn_fp16_1x_coco/panoptic_fpn_r2_50_fpn_fp16_1x_coco_20221114_224729.log.json
- https://download.openmmlab.com/mmdetection/v2.0/rfnext/panoptic_fpn_r2_50_fpn_fp16_1x_coco/panoptic_fpn_r2_50_fpn_fp16_1x_coco-fa6c51f0.pth

**Comment by lzyhha** [comment]
Thanks. And I have updated all links.

**Review by zytx121 — APPROVED** [review]
LGTM

File: configs/panoptic_fpn/metafile.yml [hunk 18]
```diff
@@ -68,3 +68,16 @@ Models:
...
+    Config: configs/panoptic_fpn/panoptic_fpn_r2_50_fpn_fp16_1x_coco.py
+    Metadata:
+      Training Memory (GB): 3.5
+      Epochs: 12
+    Results:
+    - Task: Panoptic Segmentation
+      Dataset: COCO
+      Metrics:
+        PQ: 42.5
+    Weights: weights
```
Comment (jbwang1997) [inline configs/panoptic_fpn/metafile.yml:83]: Please change the weight link.

File: configs/rfnext/metafile.yml [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+    Results:
+      - Task: Object Detection
+        Dataset: COCO
+        Metrics:
+          box AP: 39.3
+      - Task: Instance Segmentation
+        Dataset: COCO
+        Metrics:
+          mask AP: 37.1
+    Weights: weights
```
Comment (jbwang1997) [inline configs/rfnext/metafile.yml:181]: Please remember to update the link in metafile.yml

**Review by jbwang1997 — APPROVED** [review]
LGTM
